Add memory caching into pipeline

weixuanfu opened the pull request:
## What does this PR do?

Add `Memory` caching to `sklearn.pipeline.Pipeline` for saving computational time on refitting time-consuming transformers, especially in `TPOT MDR` mode 

## Where should the reviewer start?

fit() in base.py

## How should this PR be tested?

```
from tpot import TPOTClassifier
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from time import time
import numpy as np
TARGET_NAME = 'class'
INPUT_SEPARATOR = '\t'
DATASET_NAME = 'GAMETES_Epistasis_2-Way_1000atts_0.4H_EDM-1_EDM-1_1'
inputfile = 'https://github.com/EpistasisLab/penn-ml-benchmarks/raw/master/datasets/{DATASET_NAME}/{DATASET_NAME}.csv.gz'.format(DATASET_NAME=DATASET_NAME)

input_data = np.recfromcsv(
            inputfile,
            delimiter=INPUT_SEPARATOR,
            dtype=np.float64,
            case_sensitive=True
        )

features = np.delete(
        input_data.view(np.float64).reshape(input_data.size, -1),
        input_data.dtype.names.index(TARGET_NAME),
        axis=1
    )

training_features, testing_features, training_target, testing_target = train_test_split(features, input_data[TARGET_NAME], random_state=42)

start_time = time()
tpot = TPOTClassifier(generations=5, population_size=50, verbosity=2, random_state=40, config_dict="TPOT MDR")
tpot.fit(training_features, training_target)
print(tpot.score(testing_features, testing_target))
print('Time used', time()-start_time)

```


## What are the relevant issues?

#577 


## Questions:

- Do the docs need to be updated? No
- Does this PR add new (Python) dependencies? No

Commit:
use Memory to caching pipeline

Commit:
reset memory to none

rhiever commented on tpot/base.py:
```diff
@@ -238,6 +240,7 @@ def __init__(self, generations=100, population_size=100, offspring_size=None,
         self.early_stop = early_stop
         self._last_optimized_pareto_front = None
         self._last_optimized_pareto_front_n_gens = 0
+        self.memory = None
```
Let's make `memory` another TPOT parameter, with the following options:

* `None` (default): No memory
* `'auto'`: Memory with a temp directory (TPOT creates the temp directory and cleans it up upon shutdown)
* directory path string: Memory with the provided directory (TPOT does NOT clean the directory up upon shutdown)
* `Memory` object: Memory with provided `Memory` object (TPOT does NOT clean the directory up upon shutdown)

weixuanfu replied:
Make sense, will do

rhiever replied:
In the docs for these options, we should clarify that we allow the user to specify a custom directory path or `Memory` object in case they want to re-use the memory cache in future TPOT runs (or a `warm_start` run). However, they shouldn't expect the memory cache to be meaningfully interpretable in any way.

We should also make it clear that if they provide a custom directory path or `Memory` object, it is their responsibility to clean up the memory cache when they're done with it.

rhiever commented:
Even if the speedup for regular TPOT is fairly minor with this `memory` option, I figure it doesn't hurt to add it as a new parameter.

In this PR, let's make sure to:

* Update the docstrings
* Update the docs (both API and a new `Using` section on how to use the new `memory` option)
* Add a few unit tests

Commit:
add memory as parameter

Commit:
fix a bug and add a unit test

Commit:
add 5 more unit tests

Commit:
fix inital value of self._memory

Commit:
update driver

Commit:
fix unit  test in driver

Commit:
unknow issue in driver test

Commit:
invisable space

Commit:
update OUTPUT_FILE common line mode for better unit tests

Commit:
upate api.md

Commit:
update using.md for command line mode

Commit:
add a new section and examples for enabling memory

weixuanfu commented:
Update:
Added `memory` parameter in TPOTbase
Added 6 unit tests for this function
Also added `memory` option into CLI
Updated docs in both `api.md` and `using.md`

Commit:
clean docs

rhiever commented on tests/tpot_tests.py:
```diff
@@ -684,6 +687,116 @@ def test_fit_4():
+        offspring_size=2,
+        generations=1,
+        config_dict='TPOT light',
+        memory="auto",
```
Single quotes.

weixuanfu replied:
Fixed in the commit below

rhiever commented on docs_sources/using.md:
```diff
@@ -504,6 +516,29 @@ For more detailed examples of how to customize TPOT's operator configuration, se
 
 Note that you must have all of the corresponding packages for the operators installed on your computer, otherwise TPOT will not be able to use them. For example, if XGBoost is not installed on your computer, then TPOT will simply not import nor use XGBoost in the pipelines it considers.
 
+# Enabling memory caching in TPOT
```
Maybe change section title to "Pipeline caching in TPOT"

weixuanfu replied:
Fixed in the commit below

rhiever commented on tpot/base.py:
```diff
@@ -238,6 +252,8 @@ def __init__(self, generations=100, population_size=100, offspring_size=None,
         self._last_optimized_pareto_front = None
         self._last_optimized_pareto_front_n_gens = 0
+        self.memory = memory
+        self._memory = None # initial Memory setting for sklearn pipeline
```
Why is there both `memory` and `_memory` attributes? Shouldn't there only be one?

weixuanfu replied:
`self.memory` is parameter but `_memory` can only be Memory object. `self.memory` is needed for checking the initial parameter in the end of optimization process (if it is "auto", then clean up memory caching and temporary directory).

rhiever replied:
Hmm, good point. I suppose `fit` shouldn't override that parameter value.

Commit:
update docs and clean codes

Commit:
update docs

Commit:
clean codes

Commit:
add some clean up in unit test for Memory

Commit:
quick PolynomialFeatures test

rhiever commented on docs_sources/api.md:
```diff
@@ -147,6 +148,19 @@ Flag indicating whether the TPOT instance will reuse the population from previou
 
+<strong>memory</strong>: a sklearn.external.joblib.Memory object or string, optional (default=None)
+<blockquote>
+If supplied, TPOT uses memory caching while optimizing.
```
Make it clear that this is pipeline caching and link to http://scikit-learn.org/stable/modules/pipeline.html#caching-transformers-avoid-repeated-computation

rhiever commented on docs_sources/api.md:
```diff
@@ -602,6 +617,19 @@ Flag indicating whether the TPOT instance will reuse the population from previou
 
+<strong>memory</strong>: a sklearn.external.joblib.Memory object or string, optional (default=None)
+<blockquote>
+If supplied, TPOT uses memory caching while optimizing.
```
Make it clear that this is pipeline caching and link to http://scikit-learn.org/stable/modules/pipeline.html#caching-transformers-avoid-repeated-computation

rhiever commented on docs_sources/using.md:
```diff
@@ -280,6 +280,18 @@ See the <a href="../using/#built-in-tpot-configurations">built-in configurations
+<td>Memory caching mode in TPOT:
+<br /><br />
+<ul>
+<li>Path for a caching directory: TPOT uses memory caching with the provided directory and TPOT does NOT clean the caching directory up upon shutdown.</li>
```
Make it clear that this is pipeline caching.

rhiever commented on docs_sources/using.md:
```diff
@@ -504,6 +516,29 @@ For more detailed examples of how to customize TPOT's operator configuration, se
 Note that you must have all of the corresponding packages for the operators installed on your computer, otherwise TPOT will not be able to use them. For example, if XGBoost is not installed on your computer, then TPOT will simply not import nor use XGBoost in the pipelines it considers.
 
+# Pipeline caching in TPOT
+With `memory` parameter, TPOT allows users to specify a custom directory path or [`sklearn.external.joblib.Memory`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/externals/joblib/memory.py#L847) in case they want to re-use the memory cache in future TPOT runs (or a `warm_start` run) even if the speedup for regular TPOT is fairly minor with this option.
```
Make it clear that this is pipeline caching.

rhiever commented on docs_sources/using.md:
```diff
@@ -504,6 +516,29 @@ For more detailed examples of how to customize TPOT's operator configuration, se
+# Method 3, with a Memory object
+cachedir = mkdtemp() # Create a temporary folder
+memory = Memory(cachedir=cachedir, verbose=0)
+tpot = TPOTClassifier(memory=memory)
```
Also show in this example how to delete the temporary directory.

rhiever commented on docs_sources/using.md:
````diff
@@ -504,6 +516,29 @@ For more detailed examples of how to customize TPOT's operator configuration, se
+tpot = TPOTClassifier(memory=memory)
+```
+
+**Note: TPOT does NOT clean up memory caches if users set a custom directory path or Memory object. Please clean up the memory caches when these caches are not necessary.**
````
Change from "Please clean up" to "We recommend that you"

rhiever commented on tpot/base.py:
```diff
@@ -190,6 +193,17 @@ def __init__(self, generations=100, population_size=100, offspring_size=None,
             Flag indicating whether the TPOT instance will reuse the population from
             previous calls to fit().
+        memory: a Memory object or string, optional (default: None)
+            String 'auto':
```
Add a line explaining what the `memory` parameter is here. Make it clear that this is pipeline caching.

rhiever commented on tpot/driver.py:
```diff
@@ -351,6 +351,18 @@ def _get_arg_parser():
+        default=None,
+        type=str,
+        help=(
+            'Path of a directory for memory caching or \"auto\" for using temporary '
```
Clarify that this is pipeline caching.

weixuanfu replied:
OK, docs were added in the commit below.

Commit:
Clarify that this is pipeline caching.

Commit:
clean cdoes

rhiever committed:
Update using.md

rhiever committed:
Update using.md

rhiever committed:
Update driver.py

rhiever committed:
Update base.py
